### PR TITLE
LibGL+LibSoftGPU: Implement shaders

### PR DESCRIPTION
**Pull request by supercomputer7**
This is an attempt to revive from the work that has been done in #15272.

It had a couple of conflicts because of changes in LibGL recently. I still need to fix the last comments of that PR before this is mergeable :)

cc @sunverwerth

### CODE REVIEW

File: Userland/Libraries/LibGL/Shaders/Shader.h
Comment (sunverwerth) [inline Userland/Libraries/LibGL/Shaders/Shader.h:31]: Was this added in my original PR? Looks weird, maybe returning a const reference would be nicer.

File: Userland/Libraries/LibGL/Shaders/Program.cpp
Comment (GMTA) [inline Userland/Libraries/LibGL/Shaders/Program.cpp:27]: We don't need these `break`s after a `return`.

File: Tests/LibGL/TestShaders.cpp
Comment (GMTA) [inline Tests/LibGL/TestShaders.cpp:21]: These defines aren't used.

File: Userland/Libraries/LibGL/Shader.cpp
Comment (GMTA) [inline Userland/Libraries/LibGL/Shader.cpp:152]: This whitespace fix should go into the commit that introduced it

File: Userland/Libraries/LibSoftGPU/Device.cpp
Comment (GMTA) [inline Userland/Libraries/LibSoftGPU/Device.cpp:822]: These braces are not necessary for one-liners; they weren't there before

**Review by GMTA — CHANGES_REQUESTED** [review]

**Comment by GMTA** [comment]
Thanks @supercomputer7 ! There are also some comments in the old PR that are still relevant, but I think you found those.

File: Userland/Libraries/LibGL/Shaders/Shader.h
Comment (supercomputer7) [reply Userland/Libraries/LibGL/Shaders/Shader.h:31]: Not sure, I will fix this otherwise.

Comment (supercomputer7) [reply Userland/Libraries/LibGL/Shaders/Shader.h:31]: On second thought, I don't think we need to change this. Everything else that uses this function assumes it will get a pointer, and it's not guaranteed that you will have a valid reference to the `ObjectFile`.

Comment (sunverwerth) [reply Userland/Libraries/LibGL/Shaders/Shader.h:31]: Right, I think I did it like that because of the `ObjectFile` vector in `GL::Program::link()`. I remember this was done on purpose.

File: Userland/Libraries/LibGL/Shader.cpp
Comment (sunverwerth) [inline Userland/Libraries/LibGL/Shader.cpp:51]:
```suggestion
        if (length == nullptr || length[i] < 0)
```

Copying @Quaker762 's suggestion from the old PR.

**Review by sunverwerth — CHANGES_REQUESTED** [review]

**Comment by supercomputer7** [comment]
This should be ready for merging to the tree now.

**Review by sunverwerth — APPROVED** [review]
LGTM, but since I was the original author maybe someone else should have the final say :^) @GMTA ?

**Comment by supercomputer7** [comment]
Seems like CI failed because of network errors. So I think it's still mergeable.

**Comment by sunverwerth** [comment]
I tested a few of our 3d apps on this branch. Everything still seems to be in order.
Screenshot at 2022-12-04 14-49-29
Screenshot at 2022-12-04 15-00-14
Screenshot at 2022-12-04 15-03-42
Screenshot at 2022-12-04 15-06-55

**Comment by sunverwerth** [comment]
Reminds me that we badly need more automated tests for LibGL :^)

File: Userland/Libraries/LibGL/Shaders/Shader.h
Comment (ADKaster) [inline Userland/Libraries/LibGL/Shaders/Shader.h:27]: This code needs some help regarding the String --> DeprecatedString refactor. At least m_sources and m_info_log here should be converted to DeprecatedString entries until we can refactor later.

**Comment by supercomputer7** [comment]
I refactored this to use the new String class. While I looked through the code, I found places where the code could return an `Error` in a function that has `ErrorOr<void>` return type but some parts of the function neglected it, so I fixed these few cases as well.
I consider this ready to merge as soon as CI is green and @sunverwerth puts his approval on this.

File: Userland/Libraries/LibGL/Shaders/Shader.cpp
Comment (sunverwerth) [inline Userland/Libraries/LibGL/Shaders/Shader.cpp:20]: Why is this called `source_code_name`?

**Review by sunverwerth — APPROVED** [review]

**Review by ADKaster — APPROVED** [review]

**Pull request merged**
